kuri_joystick_teleop/src/kuri_joystick_teleop.cpp

```diff
@@ -0,0 +1,392 @@
+        Left Joystick Push, Right Joystick Push,
+      ]
+
+      For this mapping:
+      - Left Joystick L-R will be angular velocity
+      - Right Joystick U-D will be linear velocity
+      - LB and LR will be e-stop buttons
+      - Start will be a re-enable button
+      - Arrow Keys U-D head tilt
+      - Arrow Keys L-R head pan
+      - Button A will toggle the eyes open/close
+      - Button X centers the head (pan)
+
+    */
+
+    if (msg.buttons[4] || msg.buttons[5]) { // if any of the estop buttons are pressed
+      std::unique_lock<std::mutex> emergency_stop_lock(emergency_stop_mutex);
+      if (!emergency_stop) new_emergency_stop = true;
+      emergency_stop = true;
+      emergency_stop_lock.unlock();
+      ROS_INFO("Joystick callback got e-stop");
+      return;
+    } else if (msg.buttons[7]) { // re-enable robot
+      std::unique_lock<std::mutex> emergency_stop_lock(emergency_stop_mutex);
+      if (emergency_stop) new_emergency_stop = true;
+      emergency_stop = false;
+      emergency_stop_lock.unlock();
+      ROS_INFO("Joystick callback got re-enable");
+      return;
+    }
+
+    // NOTE: this was an attempt to use just one 2DOF joystick for both linear
+    // and rotation movements, but turned out to be unintuitive.
+
+    /* double left_right = msg.axes[0];
+    double up_down = msg.axes[1];
+    double linear_vel = 0.0, angular_vel = 0.0;
+
+    if (left_right != 0.0 || up_down != 0.0) {
+      double magnitude = std::sqrt(std::pow(left_right, 2.0) + std::pow(up_down, 2.0));
+
+      if (std::abs(up_down) >= std::abs(left_right)) {
+        linear_vel = std::copysign(magnitude, up_down);
+        angular_vel = std::copysign(magnitude, left_right)*std::abs(left_right)/std::abs(up_down);
+      } else {
+        linear_vel = std::copysign(magnitude, up_down)*std::abs(up_down)/std::abs(left_right);
+        angular_vel = std::copysign(magnitude, left_right);
+      }
+    } */
+
+    double linear_vel = msg.axes[4];
+    double angular_vel = msg.axes[0];
+
+    geometry_msgs::Twist vel;
+    vel.linear.x = linear_max * linear_vel;
+    vel.linear.y = 0;
+    vel.linear.z = 0;
+    vel.angular.x = 0;
+    vel.angular.y = 0;
+    vel.angular.z = angular_max * angular_vel;
+    std::unique_lock<std::mutex> joystick_vel_lock(joystick_vel_mutex);
+    joystick_vel = vel;
+    new_joystick_vel = true;
+    joystick_vel_lock.unlock();
+
+    int head_tilt = msg.axes[7];
+    int head_pan = msg.axes[6];
+    std::unique_lock<std::mutex> head_state_command_lock(head_state_command_mutex);
+    head_pan_reset = msg.buttons[2];
+    head_tilt_command = head_tilt;
+    head_pan_command = head_pan;
+    new_head_state_command = true;
+    head_state_command_lock.unlock();
+
+    int eye_toggle = msg.buttons[0];
+    std::unique_lock<std::mutex> eye_state_command_lock(eye_state_command_mutex);
+    new_eye_state_command = (eye_toggle == 1.0);
+    eye_state_command_lock.unlock();
```

Comment:
Hmm, I am okay with making them scoped to prevent people who may make future edits to the code from messing up. But I strongly prefer explicitly calling unlock as opposed to letting the destructor implicitly unlock it (I think it improves readability, especially for people who may be familiar with locks but not with the C++ magic that unlocks it). So I would basically be adding curly brackets around the constructor and unlock call here -- does that address your concern?

Reply:
Since it's already a scoped lock, it's only a style thing, but [it's something you'll never see in good practice](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#e6-use-raii-to-prevent-leaks). Manually unlocking things is a code smell ([textbook bug source](https://en.cppreference.com/w/cpp/language/raii)), so it has the effect of making the code less readable for people who are read into the language. If you prefer to leave them, then I would just let them be without changes, but I guarantee you'll raise an eyebrow when you reread the code in a few years :stuck_out_tongue_winking_eye:

Reply:
But what do you think about something like lines 175-192? Because I am manually calling unlock, I can have it unlock at different times in the if and else clauses, ensuring that the lock is only held for the minimum amount of time that we need the shared resources (in this case, the shared resource is the emergency_stop variable). If I were using RAII, the lock would be held until the whole if-statement has finished, which means it would be held for longer than necessary, which may slow down other threads. A similar pattern can be seen in lines 246-259.

Is there an RAII best-practice for such cases? I don't see a particularly easy/readable way to avoid manually calling unlock.